### PR TITLE
First working pipeline for training and evaluating lightweight Cache strategy

### PR DESCRIPTION
**Pull request by StefanieSwz**
I'll do the PR, so that I can set a few comments for you to have a look at for Code Review. Everything should be commented in the code though. :)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 __pycache__
+.env
```
Comment (StefanieSwz) [inline .gitignore:2]: .env for Antropic API Key

File: cache.py [hunk 2]
```diff
@@ -108,6 +111,19 @@ def add_cache_arguments(parser: argparse.ArgumentParser):
         action="store_true",
         help="Whether to accumulate number of times a token was unimportant (binary) versus raw un-normalized probabilities. If true, more memory efficient.",
     )
+    parser.add_argument(
```
Comment (StefanieSwz) [inline cache.py:114]: Arguments for lightweight strategy

File: cache.py [hunk 3]
```diff
@@ -149,6 +165,77 @@ def create_window_attention_mask(seq_len, window_size, device, global_tokens: in
     return mask
 
 
+def save_lightweight(model: nn.Module, kwargs: Dict[str, Any], identifier: str) -> Path:
```
Comment (StefanieSwz) [inline cache.py]: Saving and loading lightweight weights are on model level, but as they are cache related the functions are defined here

File: generation_utils.py [hunk 4]
```diff
@@ -384,7 +385,10 @@ def setup_caches(
 
     with torch.device(device):
         model.setup_caches(max_batch_size=1, **cache_kwargs)
-
+        if cache_kwargs["cache_strategy"][0] == "lightweight":
```
Comment (StefanieSwz) [inline generation_utils.py:394]: load trained weights after cache setup

File: cache.py [hunk 5]
```diff
@@ -149,6 +165,77 @@ def create_window_attention_mask(seq_len, window_size, device, global_tokens: in
...
+
+    Returns:
+        None
+    """
+    trained_weights: Dict[str, torch.Tensor] = torch.load(
+        checkpoint_path, map_location="cpu"
+    )
+    model_state: Dict[str, torch.Tensor] = model.state_dict()
+    matched_weights = {k: v for k, v in trained_weights.items() if k in model_state}
+    unmatched_weights = set(trained_weights.keys()) - set(model_state.keys())
```
Comment (ayyoobimani) [inline cache.py]: maybe assert len(unmatched_weights) == 0

File: model.py [hunk 6]
```diff
@@ -399,16 +436,29 @@ def forward(
...
-            k_rep,
-            v_rep,
-            attn_mask=kv_mask if mask is None else mask,
-            dropout_p=0.0,
-            attn_top_k=attn_top_k,
-            # Ask the cache if needs attention scores returned (we cannot use FlexAttention if so)
-            return_attn=self.kv_cache.return_attn(),
-        )
+        if self.train_mode:
+            with torch.no_grad():
```
Comment (ayyoobimani) [inline model.py]: At inference time you would have to do torch.no_grad() anyway

File: train_lightweight.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+# Configuration
+# TODO: Support terminal arguments via argparse
+CHECKPOINT_PATH = Path("./checkpoints/Qwen/Qwen2-1.5B-Instruct/model.pth")
+TOKENIZER_PATH = Path(CHECKPOINT_PATH).parent / "tokenizer.model"
+DATASET_PATH = "HuggingFaceH4/ultrachat_200k"
+BATCH_SIZE = 1
+EPOCHS = 5
+LEARNING_RATE = 3e-6
+MAX_SEQ_LENGTH = 2048
+IGNORE_INDEX = -100
```
Comment (ayyoobimani) [inline train_lightweight.py]: gradient accumulation

File: train_lightweight.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+                    current_token_position : current_token_position + message_length
+                ] = IGNORE_INDEX
+
+            current_token_position += message_length
+
+        return {
+            "input_ids": tokenized_ids,
+            "labels": labels,
+        }
+
```
Comment (ayyoobimani) [inline train_lightweight.py:134]: Use weights and biases to visualize

File: train_lightweight.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+        them to IGNORE_INDEX.
+
+        Args:
+            example (Dict[str, Any]): A dictionary representing a chat example with a "messages" key.
+
+        Returns:
+            Dict[str, torch.Tensor]: A dictionary with keys "input_ids" and "labels" containing tokenized tensors.
+        """
+        dialog = example["messages"]
+        tokenized_ids = self.tokenizer.encode_dialog_prompt(dialog)
```
Comment (ayyoobimani) [inline train_lightweight.py:114]: whats the difference between tokenizer.apply_chat_template and encode dialog_prommpt?!

File: train_lightweight.py [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+        tokenized_ids = torch.tensor(tokenized_ids, dtype=torch.long)
+        labels = tokenized_ids.clone()
+
```
Comment (ayyoobimani) [inline train_lightweight.py:109]: One thing you could do in the future if the loss doesn't go down is to set label to -100 for user tokens and only calculate loss for assistant tokens

File: train_lightweight.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+            input_ids, input_pos, mask=causal_mask, is_prefill=True
+        )  # prefill = true for training on full sequence
+
+        # Loss computation
+        logits = logits[:, :-1].reshape(-1, logits.size(-1))
+        labels = labels.reshape(-1)
+        loss_epoch = loss(logits, labels)
+
+        # Backward pass
+        loss_epoch.backward()
```
Comment (ayyoobimani) [inline train_lightweight.py]: Maybe you you'd have to normalize loss with the sequence length!

File: train_lightweight.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+
+        optimizer.step()
+
+        total_loss += loss_epoch.item()
+
```
Comment (ayyoobimani) [inline train_lightweight.py]: do an eval loss monitoring here. Also to W n B
